### PR TITLE
[error] Add line number and source code to exception

### LINKED ISSUE #3612: Print the related source code and its position in the file when an error occurs
**Concisely describe the proposed feature**
I would like to add the feature that prints the related source code and its position in the file when an error occurs when building the AST.

**Describe the solution you'd like (if any)**
Keep a copy of the source code of the taichi kernel/function inside the context and utilize the position information in the AST node to find out the source code related to the node. When an exception is thrown, we can catch it in the transformer and add the extra information in the error message and then throw it.

### PR DESCRIPTION
**Pull request by lin-hitonami**
Related issue = close #3612

### CODE REVIEW

File: python/taichi/lang/exception.py [hunk 1]
```diff
@@ -4,3 +4,7 @@ class TaichiSyntaxError(Exception):
 
 class InvalidOperationError(Exception):
     pass
+
+
+class TaichiError(Exception):
```
Comment (strongoier) [inline python/taichi/lang/exception.py:9]: What about renaming this to `TaichiCompilationError`? `TaichiError` sounds a bit weird.

**Review by strongoier — COMMENTED** [review]
Looks great! I feel like the modifications in #3439 is no longer needed now. Do I understand correctly?

**Comment by lin-hitonami** [comment]
> Looks great! I feel like the modifications in #3439 is no longer needed now. Do I understand correctly?

We can define some exceptions in pybind11 ffi part so that the customized exceptions in C++ can be converted to python exceptions naturally.

**Review by strongoier — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**